File: artha/src/lib/stock/presentation.ts
import { hasDataValue } from "@/lib/utils/formatters";

export type DataStatus = "live" | "delayed" | "partial" | "unavailable";

export interface DataMeta {
  asOf: string | null;
  status: DataStatus;
  coverage: number;
  note?: string | null;
  unitLabel?: string | null;
}

export function clampCoverage(value: number): number {
  if (Number.isNaN(value) || !Number.isFinite(value)) return 0;
  return Math.max(0, Math.min(1, value));
}

export function getCoverage(values: unknown[]): number {
  if (values.length === 0) return 0;
  const populated = values.filter(hasDataValue).length;
  return clampCoverage(populated / values.length);
}

export function getObjectCoverage<T extends Record<string, unknown>>(
  record: T | null | undefined,
  keys: Array<keyof T>,
): number {
  if (!record) return 0;
  return getCoverage(keys.map((key) => record[key]));
}

export function getDatasetCoverage<T>(
  rows: T[] | null | undefined,
  selectors: Array<(row: T) => unknown>,
): number {
  if (!rows || rows.length === 0 || selectors.length === 0) return 0;

  let total = 0;
  let populated = 0;
  for (const row of rows) {
    for (const selector of selectors) {
      total += 1;
      if (hasDataValue(selector(row))) populated += 1;
    }
  }

  return total === 0 ? 0 : clampCoverage(populated / total);
}

export function getLatestDate(candidates: Array<string | null | undefined>): string | null {
  const parsed = candidates
    .map((value) => {
      if (!value) return null;
      const timestamp = Date.parse(value);
      return Number.isNaN(timestamp) ? null : { value, timestamp };
    })
    .filter((entry): entry is { value: string; timestamp: number } => entry != null);

  if (parsed.length === 0) {
    return candidates.find((candidate): candidate is string => Boolean(candidate)) ?? null;
  }

  parsed.sort((left, right) => right.timestamp - left.timestamp);
  return parsed[0]?.value ?? null;
}

export function buildDataMeta(input: {
  asOf?: string | null;
  asOfCandidates?: Array<string | null | undefined>;
  coverage?: number;
  note?: string | null;
  status?: DataStatus;
  unitLabel?: string | null;
}): DataMeta {
  const asOf = input.asOf ?? getLatestDate(input.asOfCandidates ?? []);
  const coverage = clampCoverage(input.coverage ?? 0);
  const status = input.status
    ?? (coverage <= 0 ? "unavailable" : coverage < 0.999 ? "partial" : "delayed");

  return {
    asOf,
    status,
    coverage,
    note: input.note ?? null,
    unitLabel: input.unitLabel ?? null,
  };
}

export function shouldRenderChart(points: unknown[] | null | undefined, minPoints = 3): boolean {
  return (points?.length ?? 0) >= minPoints;
}

export function shouldCollapseSparseGrid(
  values: unknown[],
  threshold = 0.6,
): boolean {
  return getCoverage(values) < threshold;
}

export function completenessScore(values: unknown[]): number {
  return getCoverage(values);
}

export function formatCoverageLabel(coverage: number): string {
  return `${Math.round(clampCoverage(coverage) * 100)}% coverage`;
}

export function dedupeByKey<T>(items: T[], keySelector: (item: T) => string): T[] {
  const seen = new Set<string>();
  const result: T[] = [];

  for (const item of items) {
    const key = keySelector(item);
    if (!key || seen.has(key)) continue;
    seen.add(key);
    result.push(item);
  }

  return result;
}

File: artha/services/artha-api/core/presentation.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Callable


def clamp_coverage(value: float) -> float:
    if value != value or value == float("inf") or value == float("-inf"):
        return 0.0
    return max(0.0, min(1.0, value))


def has_data_value(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return value.strip() != ""
    if isinstance(value, (list, tuple, dict, set)):
        return len(value) > 0
    return True


def get_coverage(values: list[Any]) -> float:
    if not values:
        return 0.0
    populated = sum(1 for value in values if has_data_value(value))
    return clamp_coverage(populated / len(values))


def get_object_coverage(record: dict[str, Any] | None, keys: list[str]) -> float:
    if not record:
        return 0.0
    return get_coverage([record.get(key) for key in keys])


def get_dataset_coverage(rows: list[Any] | None, selectors: list[Callable[[Any], Any]]) -> float:
    if not rows or not selectors:
        return 0.0
    total = 0
    populated = 0
    for row in rows:
        for selector in selectors:
            total += 1
            if has_data_value(selector(row)):
                populated += 1
    return clamp_coverage(populated / total) if total else 0.0


def get_latest_date(candidates: list[str | None]) -> str | None:
    parsed: list[tuple[str, float]] = []
    for value in candidates:
        if not value:
            continue
        try:
            parsed.append((value, datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp()))
        except ValueError:
            continue
    if not parsed:
        for candidate in candidates:
            if candidate:
                return candidate
        return None
    parsed.sort(key=lambda item: item[1], reverse=True)
    return parsed[0][0]


def build_data_meta(input_data: dict[str, Any]) -> dict[str, Any]:
    as_of = input_data.get("asOf") or get_latest_date(input_data.get("asOfCandidates", []))
    coverage = clamp_coverage(float(input_data.get("coverage", 0.0)))
    status = input_data.get("status") or ("unavailable" if coverage <= 0 else "partial" if coverage < 0.999 else "delayed")
    return {
        "asOf": as_of,
        "status": status,
        "coverage": coverage,
        "note": input_data.get("note"),
        "unitLabel": input_data.get("unitLabel"),
    }


def compute_coverage(latest_date: datetime | None) -> dict[str, Any]:
    if latest_date is None:
        return {"status": "unavailable", "coverage_pct": 0.0, "as_of": None}

    now = datetime.now(UTC)
    reference = latest_date if latest_date.tzinfo else latest_date.replace(tzinfo=UTC)
    staleness = max((now - reference).days, 0)

    if staleness <= 1:
        status = "live"
    elif staleness <= 7:
        status = "delayed"
    elif staleness <= 90:
        status = "partial"
    else:
        status = "unavailable"

    if staleness <= 7:
        coverage_pct = 100.0
    elif staleness > 90:
        coverage_pct = 0.0
    else:
        span = 90 - 7
        remaining = 90 - staleness
        coverage_pct = max(0.0, round((remaining / span) * 100, 2))

    return {"status": status, "coverage_pct": coverage_pct, "as_of": reference.isoformat()}


def annotate_meta(rows: list[dict[str, Any]], date_field: str) -> dict[str, Any]:
    latest: datetime | None = None
    for row in rows:
        raw = row.get(date_field)
        if raw is None:
            continue
        if isinstance(raw, datetime):
            candidate = raw
        else:
            try:
                candidate = datetime.fromisoformat(str(raw).replace("Z", "+00:00"))
            except ValueError:
                continue
        if latest is None or candidate > latest:
            latest = candidate
    return compute_coverage(latest)
